### DISCUSSION TITLE
Attach File(s) To Test Result Of a Test Case

### DISCUSSION THREAD

**Discussion by bormm** [pester/Pester#2148] [Ideas]
We are using Pester to do simple tests of our complex software product accessed using a API. Sometimes getting the failed is not enough to have any idea what happened and why. NUnit allows attaching files to the test result, which for example would allow attaching log files. Would allow attaching the result that was not expected and so one. The feature is extremely handy to use and helpful, as you have a the diagnostic data where they belong: at the failed test.
NUnit XML Format: https://docs.nunit.org/articles/nunit/technical-notes/usage/Test-Result-XML-Format.html#attachments
NUnit API: https://docs.nunit.org/articles/nunit/writing-tests/TestContext.html#addtestattachment-37

Not exactly the same, but related: In NUnit it is also possible to add just additional logging. I also don't see anything like this in Pester: https://docs.nunit.org/articles/nunit/writing-tests/TestContext.html#writeline

In Pester its only possible to set the test result explicit using [Set-ItResult](https://pester.dev/docs/commands/Set-ItResult) but I think that could be a good starting point for something like **Add-ItFileAttachment** ?

**Comment by fflaten** [comment]
Thanks for feedback!

Something similar is suggested in #2039 . Would https://github.com/pester/Pester/issues/2039#issuecomment-886114682 work for you?

The JUnit schema doesn't have a similar element, however output will be available in the future NUnit3 support.

**Reply by bormm** [reply]
Thank you @fflaten . I searched a while but haven't found the mentioned ticket as none of my keywords matched the description.
I don't see how the linked workaround would help me or the creator of #2039, but we can continue the discussion there if you want.

**Reply by fflaten** [reply]
I read the NUnit format as  testsuite/-case -> attachments->attachment-> filePath/description which is only a path-reference. The external files would not be embedded in the xml-report itself. 

As a workaround, that reference could just as well be included in the output, couldn't it? I understand it's not a direct replacement in case you have tools etc that look for the NUnit attachment-nodes.

**Reply by bormm** [reply]
Exactly, that is what I mean: The tools will not do anything with that text, unless some tool detects filenames and does something.
Just imagine what is currently working with all pieces in places: We already add for example images to our test result where we use NUnit. Like some difference images, expected image and so one and in the CI/CD or Visual Studio you can just look the image attached to "image looks different than expected" directly in the test browser. That's really cool.

So this text would be a good hint to don't miss checking addition artifact files for some logs or images, but its really primitive comparing what the real thing does.

**Reply by nohwnd** [reply]
Commented on your comment in the other issue. We don't upload files from Pester right now. But azdo makes it relatively easy: https://docs.microsoft.com/en-us/azure/devops/pipelines/scripts/logging-commands?view=azure-devops&tabs=bash#addattachment-attach-a-file-to-the-build

**Reply by nohwnd** [reply]
You still need to make the connection between a test case and a file. There are many stores in azdo though, and there is attachments page on testcase, but IDK if that is usable only through TRX output, which we don't support.
